source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,617 @@
+        ShadowByte = kUsmDeviceRedzoneMagic;
+        break;
+    case USMMemoryType::SHARE:
+        ShadowByte = kUsmSharedRedzoneMagic;
+        break;
+    case USMMemoryType::MEM_BUFFER:
+        ShadowByte = kMemBufferRedzoneMagic;
+        break;
+    default:
+        ShadowByte = 0xff;
+        assert(false && "Unknow AllocInfo Type");
+    }
+
+    // Left red zone
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->AllocBegin,
+                                AllocInfo->UserBegin - AllocInfo->AllocBegin,
+                                ShadowByte, LastEvent, &LastEvent));
+
+    // Right red zone
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, TailBegin,
+                                TailEnd - TailBegin, ShadowByte, LastEvent,
+                                &LastEvent));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::updateShadowMemory(ur_queue_handle_t Queue) {
+    auto Context = getContext(Queue);
+    auto Device = getDevice(Queue);
+    assert(Device != nullptr);
+
+    auto ContextInfo = getContextInfo(Context);
+
+    auto HostInfo = ContextInfo->getDeviceInfo(nullptr);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::shared_lock<ur_shared_mutex> HostGuard(HostInfo->Mutex,
+                                                std::defer_lock);
+    std::unique_lock<ur_shared_mutex> DeviceGuard(DeviceInfo->Mutex,
+                                                  std::defer_lock);
+    std::scoped_lock<std::shared_lock<ur_shared_mutex>,
+                     std::unique_lock<ur_shared_mutex>, ur_mutex>
+        Guard(HostGuard, DeviceGuard, QueueInfo->Mutex);
+
+    ur_event_handle_t LastEvent = QueueInfo->LastEvent;
+
+    // FIXME: Always update host USM, but it'd be better to update host USM
+    // selectively, or each devices once
+    for (auto &AllocInfo : HostInfo->AllocInfos) {
+        UR_CALL(enqueueAllocInfo(Context, Device, Queue, AllocInfo, LastEvent));
+    }
+
+    for (auto &AllocInfo : DeviceInfo->AllocInfos) {
+        UR_CALL(enqueueAllocInfo(Context, Device, Queue, AllocInfo, LastEvent));
+    }
+    DeviceInfo->AllocInfos.clear();
+
+    QueueInfo->LastEvent = LastEvent;
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::addContext(ur_context_handle_t Context) {
+    auto ContextInfo = std::make_shared<ur_sanitizer_layer::ContextInfo>();
+
+    // Host Device
+    auto DeviceInfo = std::make_shared<ur_sanitizer_layer::DeviceInfo>();
+    DeviceInfo->Type = DeviceType::CPU;
+    DeviceInfo->Alignment = ASAN_SHADOW_GRANULARITY;
+
+    // TODO: Check if host asan is enabled
+    DeviceInfo->ShadowOffset = 0;
+    DeviceInfo->ShadowOffsetEnd = 0;
+
+    ContextInfo->DeviceMap.emplace(nullptr, std::move(DeviceInfo));
+
+    std::scoped_lock<ur_shared_mutex> Guard(m_ContextMapMutex);
+    m_ContextMap.emplace(Context, std::move(ContextInfo));
```

Comment:
what if Context is emplaced twice? Is an assert needed here?

Reply:
Yep, I'll add an assert here.